Feat!: Add support for multiple virtual layers controlled by gateway

themisvaltinos opened the pull request:
This update adds support for managing a virtual layer per gateway, so a `gateway` when specified in the model will be responsible to create the views of the virtual layer as well as the physical tables. This is controlled by setting the provisionally called `gateway_managed_virtual_layer` flag to true in the config.

```yaml
gateways:
  local:
    connection:
      type: postgres
      ...

  bigquery_gw:
    connection:
      type: bigquery
    variables:
      overriden_var: 'bigquery'

  snowflake_gw:
    connection:
      type: snowflake
      database: db1
      ...
    variables:
      overriden_var: 'snowflake'

default_gateway: postgres

gateway_managed_virtual_layer: true
```

In this case the virtual layer will be created by the specific gateway instead of using the default gateway, which means that the catalog doesn't need to be shared between the gateways. Also the gateway specific catalog will be set for each corresponding model and the gateway specific variables will take precedence over global ones for the gateway-specified in the model.

The default behaviour remains unchanged when the flag is set to false or omitted from the config.

izeigerman commented on sqlmesh/core/environment.py:
```diff
@@ -32,12 +32,15 @@ class EnvironmentNamingInfo(PydanticModel):
     suffix_target: EnvironmentSuffixTarget = Field(default=EnvironmentSuffixTarget.SCHEMA)
     catalog_name_override: t.Optional[str] = None
     normalize_name: bool = True
+    gateway_managed_virtual_layer: bool = False
```
In the context of environment we already know that it concerns the virtual layer, so we can call this just `gateway_managed`

izeigerman replied:
or `model_gateway_managed` for clarity

izeigerman replied:
Plus `model_gateway` too to simplify `cleanup_expired_views`

themisvaltinos replied:
yes makes sense revised it

izeigerman commented on sqlmesh/core/environment.py:
```diff
@@ -49,6 +52,11 @@ def _sanitize_name(cls, v: str) -> str:
 
+    @field_validator("gateway_managed_virtual_layer", mode="before")
+    @classmethod
+    def _validate_gateway_managed_virtual_layer(cls, v: t.Any) -> bool:
```
We should reuse the `_validate_normalize_name` validator and just rename it

themisvaltinos replied:
refactored it into one validator

izeigerman commented on sqlmesh/core/snapshot/evaluator.py:
```diff
@@ -953,7 +963,12 @@ def _demote_snapshot(
+        adapter = (
+            self._get_adapter(snapshot.model_gateway)
+            if environment_naming_info.gateway_managed_virtual_layer
+            and isinstance(snapshot, Snapshot)
```
I don't think I like this approach. Let's not implicitly change the behavior depending on which type was passed, it never ends well. I'd rather extend `SnapshotTableInfo` and `SnapshotInfoLike` interface to support this flag.

izeigerman commented on sqlmesh/core/state_sync/common.py:
```diff
@@ -21,21 +21,36 @@
+    for engine_adapter, expired_catalog, expired_schema in {
         (
+            (
+                engine_adapter := (
```
This construct is really hard to comprehend

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -2287,22 +2290,45 @@ def _context_diff(
+                snapshots = self.state_sync.get_snapshots(environment.snapshots)
+                for snapshot_id, snapshot in snapshots.items():
+                    if snapshot.is_model and not snapshot.is_symbolic:
+                        snapshot_adapters[snapshot_id.name] = self._get_engine_adapter(
```
If we included `model_gateway` into `SnapshotTableInfo` we wouldn't have to do this.

izeigerman commented on sqlmesh/core/state_sync/common.py:
```diff
@@ -44,27 +59,40 @@ def cleanup_expired_views(
+    for engine_adapter, expired_view in {
+        (
+            (
+                engine_adapter := (
```
This looks just like 41-47 and is equally hard to comprehend

themisvaltinos replied:
revised these to be readable and more easy to understand

izeigerman commented on sqlmesh/core/state_sync/db/snapshot.py:
```diff
@@ -204,6 +204,7 @@ def delete_expired_snapshots(
         Returns:
             A tuple of expired snapshot IDs and cleanup targets.
         """
+
```
Why did this file change?

themisvaltinos replied:
yes my bad reverted it

izeigerman commented on sqlmesh/migrations/v0056_restore_table_indexes.py:
```diff
@@ -1,4 +1,4 @@
-"""Readds indexes and primary keys in case tables were restored from a backup."""
+"""Reads indexes and primary keys in case tables were restored from a backup."""
```
Why did this change?

themisvaltinos commented:
Thanks for the review @izeigerman addressed comments and updated `SnapshotTableInfo` to keep the model gateway as you suggested which simplified the code significantly, when you have time, if you can take another look

georgesittas commented on sqlmesh/core/model/definition.py:
```diff
@@ -1907,6 +1907,13 @@ def create_models_from_blueprints(
         else:
             gateway_name = None
 
+        if (
+            (catalogs := loader_kwargs.pop("catalogs", None))
+            and gateway_name
+            and (catalog := catalogs.get(gateway_name))
+        ):
+            loader_kwargs["default_catalog"] = catalog
```
Why are we popping the catalogs here?

themisvaltinos replied:
Good question, I originally had a comment here in case it was unclear so reinstated it. It is because after getting from the catalogs the corresponding catalog for this model in the case it uses the gateway property, to remove the dict so it doesn't raise Pydantic validation error since catalogs isn't a model property.

georgesittas commented on sqlmesh/core/snapshot/evaluator.py:
```diff
@@ -226,15 +226,23 @@ def promote(
-        )
+                tables.append(table)
+                if environment_naming_info.gateway_managed:
+                    table_schema = d.schema_(table.db, catalog=table.catalog)
+                    gateway_by_schema[table_schema] = snapshot.model_gateway or ""
```
Now that the catalogs don't have to be shared, can't we get a key collision in this section?

For example, let's say you have two snapshots evaluated in different engines and the corresponding table's `catalog` and `db` are the same. Is this possible now? Could it lead to problems, by e.g. overwriting a gateway?

themisvaltinos replied:
Yes, this is possible assuming both the schema and database are the same but overwriting a gateway isn't an issue since this is for creating schemas for the virtual layer. So, in this hypothetical scenario let’s say Snapshot 1’s model specifies the gateway as engine_2, while Snapshot 2’s model uses the default gateway, which is engine_1. Since they share the same database, we can assume they have access to the same catalog, so we don't have to create the same schema twice. Which engine actually performs the creation doesn't matter, since they both have access to it

izeigerman replied:
> Since they share the same database, we can assume they have access to the same catalog, 

I guess the question is what if they dont?

izeigerman replied:
Anyway, I believe we already have this problem when creating tables in the physical layer

themisvaltinos replied:
> I guess the question is what if they dont?

If they don't then there won't be a collision in this, since the table schema contains the db info as well, so they will be under a different key

themisvaltinos replied:
actually I talked to George and now I see what he meant that the unfortunate case that the catalog has identical name in two data warehouses. will revise this to cover this case

themisvaltinos commented on sqlmesh/migrations/v0078_add_gateway_managed_property.py:
yes thanks for the heads up will rename

izeigerman commented on sqlmesh/core/environment.py:
```diff
@@ -32,22 +32,27 @@ class EnvironmentNamingInfo(PydanticModel):
-        return True if v is None else bool(v)
+    def _validate_boolean_field(cls, v: t.Any, info: ValidationInfo) -> bool:
+        if v is None:
+            return info.field_name == "normalize_name"
```
Ah, now I see why they were in different validators. Sorry, I missed this before

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -2227,6 +2219,17 @@ def engine_adapters(self) -> t.Dict[str, EngineAdapter]:
         return self._engine_adapters
 
+    @cached_property
+    def catalogs(self) -> t.Dict[str, str]:
```
shouldn't this be `default_catalogs`? Or `default_catalog_per_gateway`?

themisvaltinos replied:
Yes revising this to the latter for clarity

izeigerman commented on sqlmesh/core/loader.py:
```diff
@@ -468,6 +468,7 @@ def _load() -> t.List[Model]:
                         default_catalog=self.context.default_catalog,
                         infer_names=self.config.model_naming.infer_names,
                         signal_definitions=signals,
+                        catalogs=self.context.catalogs,
```
ditto: `default_catalogs` or `default_catalogs_per_gateway`

izeigerman commented on sqlmesh/core/model/definition.py:
```diff
@@ -1907,6 +1907,14 @@ def create_models_from_blueprints(
 
+        # We pop to avoid pydantic validation issues since catalogs is not a model property
+        if (
+            (catalogs := loader_kwargs.pop("catalogs", None))
```
Instead of doing this can we just make it a first class argument?

izeigerman replied:
Also name

izeigerman commented on sqlmesh/core/plan/evaluator.py:
```diff
@@ -424,7 +424,9 @@ def _demote_snapshots(
     ) -> None:
         self.snapshot_evaluator.demote(
-            target_snapshots, environment_naming_info, on_complete=on_complete
+            target_snapshots,
```
Does this file ned to be changed?

themisvaltinos replied:
ah no it changed from other changes I reverted and then running make style so will reinstate to one line how it was

izeigerman commented on sqlmesh/core/snapshot/evaluator.py:
```diff
@@ -226,15 +226,23 @@ def promote(
-                s.qualified_view_name.table_for_environment(
-                    environment_naming_info, dialect=self.adapter.dialect
+
+        gateway_by_schema: t.Dict[t.Any, str] = {}
```
Why key is `t.Any`?

izeigerman commented on sqlmesh/core/snapshot/evaluator.py:
```diff
@@ -226,15 +226,23 @@ def promote(
+                tables.append(table)
+                if environment_naming_info.gateway_managed:
+                    table_schema = d.schema_(table.db, catalog=table.catalog)
+                    gateway_by_schema[table_schema] = snapshot.model_gateway or ""
```
Shouldn't the fallback value be `None` instead of `""`?

themisvaltinos replied:
yes we want it to be `none` thank you will change it

izeigerman commented on sqlmesh/core/state_sync/common.py:
```diff
@@ -21,21 +21,33 @@
+    adapter: EngineAdapter,
+    environments: t.List[Environment],
+    console: t.Optional[Console] = None,
+    engine_adapters: t.Optional[t.Dict[str, EngineAdapter]] = None,
```
Why is it optional? Also I think it should go before `console`.

izeigerman commented on sqlmesh/core/state_sync/common.py:
```diff
@@ -21,21 +21,33 @@
+    # Drop the schemas for the expired environments
+    for engine_adapter, expired_catalog, expired_schema in {
         (
+            (engine_adapter := get_adapter(environment.gateway_managed, snapshot.model_gateway)),
```
why do we assign twice? Shouldn't this be just
```
get_adapter(environment.gateway_managed, snapshot.model_gateway),
```?

themisvaltinos replied:
because we need the adapter before using it as well to grab the corresponding adapter's dialect:
https://github.com/TobikoData/sqlmesh/blob/c59288bb53e10211e927224e35c01f99b9cfca2f/sqlmesh/core/state_sync/common.py#L47-L52

izeigerman commented on sqlmesh/core/state_sync/db/facade.py:
```diff
@@ -199,7 +199,10 @@ def promote(
-            if not existing_environment.expired:
+            if (
+                not existing_environment.expired
+                and existing_environment.gateway_managed == environment.gateway_managed
```
I don't think this is right place for this. We need this check here instead: https://github.com/TobikoData/sqlmesh/blob/c59288bb53e10211e927224e35c01f99b9cfca2f/sqlmesh/core/state_sync/db/facade.py#L228-L232. 

The goal is to *recreate* all the views. The current block has nothing to do with that

izeigerman commented on sqlmesh/core/state_sync/common.py:
```diff
@@ -21,21 +21,33 @@
 
 def cleanup_expired_views(
-    adapter: EngineAdapter, environments: t.List[Environment], console: t.Optional[Console] = None
+    adapter: EngineAdapter,
```
This should now be `default_adapter` maybe?

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -2227,6 +2219,17 @@ def engine_adapters(self) -> t.Dict[str, EngineAdapter]:
 
+    @cached_property
+    def default_catalog_per_gateway(self) -> t.Dict[str, str]:
+        """Returns the catalogs for each engine adapter in a multi virtual layer setup when the catalog isn't shared."""
```
> in a multi virtual layer setup

wait, why is it only virtual layer? Shouldn't we apply the model gateway's default catalog to the model specifying that gateway?

izeigerman commented on sqlmesh/core/snapshot/evaluator.py:
```diff
@@ -316,24 +324,32 @@ def create(
-                for obj in objs
-            }
+            existing_objects: t.Set[str] = set()
+            for gateway, tables_by_schema in tables_by_gateway_and_schema.items():
```
Awesome. Let's leave a comment for why we're doing this gymnastics (same schema can be shared across multiple gateways)

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -2227,6 +2219,17 @@ def engine_adapters(self) -> t.Dict[str, EngineAdapter]:
+    @cached_property
+    def default_catalog_per_gateway(self) -> t.Dict[str, str]:
+        """Returns the catalogs for each engine adapter in a multi virtual layer setup when the catalog isn't shared."""
+        if self.gateway_managed_virtual_layer:
```
Shouldn't we always have default catalogs? Shouldn't this be 
```
if self._default_catalog_per_gateway is not None:
```
instead?

themisvaltinos replied:
yes I think you're right, I thought it didn't apply to projects with one virtual layer since the catalog should be the same unless overriden by the user in a model, but since this is anyway in a multi - gateway setting even without the flag it makes sense to set it correctly

izeigerman commented on sqlmesh/core/context_diff.py:
```diff
@@ -118,7 +121,7 @@ def create(
 
         create_from_env_exists = False
-        if env is None or env.expired:
+        if env is None or env.expired or env.gateway_managed != gateway_managed_virtual_layer:
```
Shouldn't we also reflect this change in https://github.com/TobikoData/sqlmesh/blob/b31c898d6834cc8147ec8a506ec2dc563ae2a733/sqlmesh/core/context_diff.py#L273? 

Otherwise, if the flag is the only thing that changed the planner won't detect a difference.

izeigerman commented on tests/core/test_integration.py:
```diff
@@ -4494,6 +4496,128 @@ def test_multi(mocker):
+    context = Context(paths=["tests/fixtures/multi_virtual_layer"])
+
+    local_db = "db.duckdb"
+    if os.path.exists(local_db):
```
To avoid problems with this state as well as concurrent test runs I suggest copying the project folder into a temporary location (use the `tmp_path` fixture)

izeigerman reviewed (approved):
very nice 👍

georgesittas reviewed (approved):
Great work!
